### ISSUE TITLE
node podresources API socket is hardcoded in kubelet

### ISSUE THREAD

**Issue by ffromani** [kubernetes/kubernetes#120625]
### What happened?

The podresources API is a kubelet-only gRPC API served by kubelet over a unix-domain socket. While the API, being gRPC, requires read-write access to the socket (request/response model), the API will only allow to inspect kubelet data and never to change the kubelet state, by design.  The socket path is hardcoded:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2902
It is possible to actually change this path, but it requires to change ALL the kubelet root directory, which has huge side effects.
It should be possible to make the path configurable.

### What did you expect to happen?

The kubelet should have an option to change the podresources API socket path and to make it fully configurable (e.g. not require changing the full kubelet root directory!)
As stretch goal, which however comes almost for free, if the argument of the option is nil/empty, the API should be disabled entirely.

### How can we reproduce it (as minimally and precisely as possible)?

run kubelet, inspect command line arguments/configuration options

### Kubernetes version

<details>

```console
$ kubectl version
any since podresources API injection
```

</details>

### Cloud provider

<details>
not relevant, applies to all the kubelet installations
</details>

### OS version

<details>
not relevant, applies to all supported variants
</details>

### Install tools

<details>
not relevant, applies to all supported variants
</details>

### Container runtime (CRI) and version (if applicable)

<details>
not relevant, applies to all supported variants
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
not relevant, applies to all supported variants
</details>

**Comment by ffromani** [comment]
/sig node
/cc @swatisehgal @SergeyKanzhelev 

we had informal chats about this topic, here's the formal bug for tracking purposes

**Comment by swatisehgal** [comment]
I can see where you are coming from and on my first read of the issue I thought it was a very good idea, almost a no-brainer but as I thought more about this I realized that this is not as harmless as it looks and we can run into a good few issue if we were to propose a change like this:

1. We have to be conscious that we have two persona at play here sys-admin/cluster operator that takes of cluster configuration and the end user who deploys the workload. The sys-admin/cluster operator here is responsible for configuring the kubelet option and would have to communicate to the end user the configured path in order to ensure that the right path is volume mounted into the pod spec. This additional communication can be an overhead and can lead to unexpected behavior. 
2. With this option, the sys-admin/cluster operator CAN configure different paths for different nodes in the cluster and the user HAS to ensure that the unix socket mounted in the client application is consistent with that configured on the node. But what if the pod is scheduled on a node where the path does not align with the configured path?? The only way this would work is if we ensure that all the nodes in the cluster is configured with the same well defined path. 
3. Also, I can understand the desire to have the ability to disable this feature if the option is empty/nil but I am concerned that this could result in some surprised/unhappy end-users just because the sys-admin forgot to update the kubelet configuration with the new option and then the workload ran into issues due to lack of the podresource API support. Because of this it might be better to default to the current hardcoded value in case the option is either not specified or empty to ensure backward compatibility. I think the ability to disable a stable feature deserves more careful evaluation  and a discussion with the wider group because it goes against the default assumptions and expectation we have as part of the graduation process of a feature.

Just stating the obvious, because we would be introducing a new kubelet option this would require a KEP and non-trivial effort from the community. The question is: do we stand to gain enough from this change to justify additional effort coupled with the risks I highlighted above? I am inclining towards a NO but I have open to an argument in favor of this change.

**Comment by ffromani** [comment]
Thanks for the feedback! you raise very good points. Let's see if we can address them.

>     1. We have to be conscious that we have two personas at play here: sys-admin/cluster operator that takes of cluster configuration and the end user who deploys the workload. The sys-admin/cluster operator here is responsible for configuring the kubelet option and would have to communicate to the end user the configured path in order to ensure that the right path is volume mounted into the pod spec. This additional communication can be an overhead and can lead to unexpected behavior.

This is very true. However it's also true that in the current state of things the actual podresources socket path is non-discoverable. The podspec owner has to discover it by side channels (reading the code?) and has to hardcode in their podspec. Ultimately, this a byproduct of the uncommon nature of the podresources API, being it node-local and exposed this way is pretty much an unicum in the ecosystem. Perhaps a better approach would be, in general, exposing the podresources socket path (and availability) in the node status? this value can't change unless kubelet reboots, so the overhead to both the node object and the apiserver would be minimal.

>     2. With this option, the sys-admin/cluster operator CAN configure different paths for different nodes in the cluster and the user HAS to ensure that the unix socket mounted in the client application is consistent with that configured on the node. But what if the pod is scheduled on a node where the path does not align with the configured path?? The only way this would work is if we ensure that all the nodes in the cluster is configured with the same well defined path.

Very true again, and would be partially addressed by exposing the api availability/path in the node status. Regarding inconsistent configuration, this is always possible and I'm afraid is a much more general (and complex) problem to solve.

>     3. Also, I can understand the desire to have the ability to disable this feature if the option is empty/nil but I am concerned that this could result in some surprised/unhappy end-users just because the sys-admin forgot to update the kubelet configuration with the new option and then the workload ran into issues due to lack of the podresource API support. Because of this, it might be better to default to the current hardcoded value in case the option is either not specified or empty to ensure backward compatibility. I think the ability to disable a stable feature deserves more careful evaluation  and a discussion with the wider group because it goes against the default assumptions and expectation we have as part of the graduation process of a feature.

This is a very fair statement. The default should be backward compatible indeed.
Having the possibility of disabling the API (albeit explicitely and opt in ) is still something we should allow: I think it should have been from the beginning.

> Just stating the obvious, because we would be introducing a new kubelet option this would require a KEP and non-trivial effort from the community.

And we will need API review

> The question is: do we stand to gain enough from this change to justify additional effort coupled with the risks I highlighted above? I am inclining towards a NO but I have open to an argument in favor of this change.

I'd be fine if we, as community, we decide to **not** pursue this effort.
I think we need to have this conversation and recorded, which is the main reason I filed this issue.

**Comment by swatisehgal** [comment]
> This is very true. However it's also true that in the current state of things the actual podresources socket path is non-discoverable. The podspec owner has to discover it by side channels (reading the code?) and has to hardcode in their podspec. 

I tend to agree about the non-discoverabilty of the path but a simple update to the documentation can address this. We already have a section on how we can [monitor resources using podresource API](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#monitoring-device-plugin-resources) which seems like the right place to document the socket path.

> Ultimately, this a byproduct of the uncommon nature of the podresources API, being it node-local and exposed this way is pretty much an unicum in the ecosystem. Perhaps a better approach would be, in general, exposing the podresources socket path (and availability) in the node status? this value can't change unless kubelet reboots, so the overhead to both the node object and the apiserver would be minimal.

Exposing this information via node status solves the problem only partially. This has somehow become a scheduling problem as we need to match the path volume mounted in the pod to the node that is able to support the configuration OR the user has to manually observe the node status and make sure the pod is scheduled to the "right" node perhaps with a `nodeSelector`. 


> Very true again, and would be partially addressed by exposing the api availability/path in the node status. Regarding inconsistent configuration, this is always possible and I'm afraid is a much more general (and complex) problem to solve.
> 

Agree that the path in the node status would solve the problem partially. The jist is that this is not as simple as it sounds on surface, all of sudden from a simple change in kubelet to make the socket path configurable we now need to make podstatus changes and perhaps changes in scheduler in order to make sure we don't cause regressions.

> This is a very fair statement. The default should be backward compatible indeed. Having the possibility of disabling the API (albeit explicitely and opt in ) is still something we should allow: I think it should have been from the beginning.

Sure, I understand and agree that disabling this feature can be useful. I am not, however convinced that coupling this with the configurable path is the right way to go.
 
> > The question is: do we stand to gain enough from this change to justify additional effort coupled with the risks I highlighted above? I am inclining towards a NO but I have open to an argument in favor of this change.
> 
> I'd be fine if we, as community, we decide to **not** pursue this effort. I think we need to have this conversation and recorded, which is the main reason I filed this issue.

Agreed!

**Comment by swatisehgal** [comment]
xref: https://github.com/kubernetes/kubernetes/issues/120625

**Comment by ffromani** [comment]
[snipping parts we reached agreement about]

> > Ultimately, this a byproduct of the uncommon nature of the podresources API, being it node-local and exposed this way is pretty much an unicum in the ecosystem. Perhaps a better approach would be, in general, exposing the podresources socket path (and availability) in the node status? this value can't change unless kubelet reboots, so the overhead to both the node object and the apiserver would be minimal.
> 
> Exposing this information via node status solves the problem only partially. This has somehow become a scheduling problem as we need to match the path volume mounted in the pod to the node that is able to support the configuration OR the user has to manually observe the node status and make sure the pod is scheduled to the "right" node perhaps with a `nodeSelector`.

I'm thinking about how to solve this problem. The controllers can patch the manifests before to submit them (e.g. to fix the path in a daemonset), but this is a side issue wrt the scheduling problem.

Thinking about it, is the same problem of resource manager configuration. We have a node property (a certain configuration leading to an extended behavior) is assumed and can't be discovered or requested.

**Comment by ffromani** [comment]
/triage accepted

still relevant, still no way forward

**Comment by ffromani** [comment]
/triage accepted

either making the path configurable or revisiting the registration mechanism, this is still an open issue.
